### DISCUSSION TITLE
How to Run Scripts in Python

### DISCUSSION THREAD

**Discussion by Kirs10-Riley** [kbuzard/labs#42] [General]
# How to Run Scripts in Python

information gathered from:

<https://realpython.com/run-python-scripts/>

<https://fangohr.github.io/blog/spyder-the-python-ide.html#:~:text=To%20execute%20the%20program%2C%20select,the%20Run%20settings%20if%20required.&text=If%20so%2C%20then%20you%20have,first%20Python%20program%20%2D%20well%20done.>

https://fangohr.github.io/blog/spyder-the-scientific-python-development-environment.html

## Terminology

**Script** - Plain text file containing Python code that is intended to be *directly executed* by the user

**Module** - plain text file containing Python code that is designed to be *imported* and used from another Python file.

**Interactive** - Python has two basic modes: script and interactive. The default mode is scripted where you write the code all at once then it can be run through a python interpreter. Interactive mode is a command line shell which gives immediate feedback for each statement or block of code.

**environment** - your work space in python

**Python Interpreter** - the program needed to run Python code and scripts

**bytecode** - translation of python code into a lower-level language that is platform-independent.

## Interpreter's Process

When running Python scripts:

1.  the interpreter will process the statements of your script in a sequential fashion
2.  then the interpreter will compile the source code to an intermediate format known as byetecode.
3.  and then a Python Virtual Machine (PVM) runs the bytecode

This process is also referred to as the Python Execution Model.

## Run a script using an IDE

When developing larger and more complex applications, it is recommended that you use an integrated development environment (IDE) or an advanced text editor.

### [Spyder - the Python IDE]( https://fangohr.github.io/blog/spyder-the-python-ide.html#:~:text=To%20execute%20the%20program%2C%20select,the%20Run%20settings%20if%20required.&text=If%20so%2C%20then%20you%20have,first%20Python%20program%20%2D%20well%20done.)

**To run scripts using spyder:** 

-  create a script in Spyder Editor pane `File` --> `New file`
- copy and paste the code inside
-   then select `Run->Run` from the menu
- the output should say `runfile('file path')`
 If this is seen then the script was successfully run. 

OR 

- Read and work through [Spyder: The Scientific Python Development Environment](https://fangohr.github.io/blog/spyder-the-scientific-python-development-environment.html)
- Skip [Section 1.4.3](https://fangohr.github.io/blog/spyder-the-scientific-python-development-environment.html#automatic-symbolic-python) if you don't have a LaTeX installation.
- In Section 1.6, replace `Run --> Configure` or by pressing `F6` with `Run --> Configuration per file` or by pressing `Ctrl + F6`
- Just read through Sections 1.8 and 1.9; come back later when you're actually writing or debugging a program to work through these sections.

> Caution when using this method is making sure your environment has all the packages needed to run the code. **See the end of this discussion for how the environment should be set up for the RamosRivera Scripts**

## Running Python Scripts Using the Command-Line/ Terminal

**To run scripts with command-line:**

-   open a command-line and type the word `python` or `python3` followed by the path to your script

> it would look something like this
>
> `$ python3 filename.py`

-   then press enter and the script should run.

to redirect the output of the script add a `> output.txt` after the file name path. This will save the output.

to add the output of consecutive executions to the end of output.txt then two angle brackets must be used. i.e. \>\> `output.txt`

> **Shortcuts/ using the script file name**
**For windows computers** - enter the name of the file containing the code and the command prompt

**To run a modules with command-line**

-   open a command-line and type the word `python` or `python3` followed by the command `-m` and the module-name

> The command would look something like this
>
> `$ python -m modulename`

## Run Python Scripts Through an Interactive Session

**Using Import**

The import command runs the code as it's final step. If the imported module does not contain functions then, there will not be any visible results. Instead the classes, function, and variables, and constants definitions will be saved to your environment. If the module contains function then the machine will generate a visual result.

To use import:

`>>> import modulename`

for this to work the Python code must be located in your current working directory and the file must be in the Python Module Search Path (PMSP).

**Using importlib and imp**

importlib is a package that contains the command `import_module()`. This command emulates an import operation.

To use:

`>>> import importlib`

`>>> importlib.import_module('modulename')`

once a module is imported, the `import` command will not run the code if used again. In this case the command `importlib.reload()` can be used. this will force the interpreter to re-import the module again and run it.

`>>> importlib.reload('modulename')`

> if using Python 2.x then the commands will be `imp` and `imp.reload()`

**Using runpy.run_module() and runpy.run_path()**

runpy is a package that contains the command `run_module()`. This command allows modules to run without importing them first.

`>>> runpy.run_module(mod_name="modulename')`

rumpy also has the command `run_path()` which allows a module to run by providing its location in the filesystem:

`>>> import runpy`

`runpy.run_path(path_name='modulename')`

the path_name parameter must be a string and can refer to the following:

-   the location of a python source file

-   the location of a compiled bytecode file

-   the value of a valid entry in the sys.path, containing a main `__main__` module (`__main__.py` file)

## Hacking exec() and execfile()

`exec()` is a built-in function that supports the dynamic execution of Python code.

`>>> exec(open('modulename').read()`

this statement open the module, reads its content, and sends it to exec(), and finally runs the code.

`execfile()` is a built-in function only available in Python 2.x that is able to run python scripts.

`>>> execfile('script.py')`

## The environment needed to run the ramosRivera scripts 

By creating this environment you are essentially downloading all the packages needed for the script to run. One important note is that these packages need to be downloaded *in order* because python packages can be dependent on one another and the script will not run properly. 

**Set up the environment: **
1) In Anaconda Command Prompt tiped:

`conda create --name Test_Jorge python=3.7` 
`conda activate Test_Jorge` 
> replace `Test_Jorge` with your desired environment name

2) Install the packages in Anaconda Command Prompt:

`conda install -c esri arcgis`
`conda install -c conda-forge mkl-service`
`conda install -c conda-forge plotnine`
`conda install -c conda-forge geopandas`
> the first two packages will finish installing after a few minutes. geopandas takes a lot longer upwards of 3+ hours. 

3) Install the modules needed in Anaconda Prompt:

`pip install opencv-python`
`pip install pytesseract`
`pip install pyap`
`pip install pdf2image`
`Pip install geopy`
`Pip install tabulate`

4) In "Anaconda Navigator"
* Go to the start menu, click environment and select "Test_Jorge" (your environment name)
* Return to the home screen and install the spyder click in Anaconda Navigator.
* Launch Spyder on your Anaconda navigator 

5) Open Spyder (your environment name), open the file on Spyder, and run the python scripts

**Comment by kbuzard** [comment]
- I transformed the links you provided into shorter hyperlinked text.
- The definition of 'interactive' is missing
   - That said, I don't think we need the "Run Python Scripts Through an Interactive Session."
      - For this project, I think you'll mainly be using Spyder, so let's put that section above the section on using the terminal
         - Please expand this part; I recommend using [this post](https://fangohr.github.io/blog/spyder-the-scientific-python-development-environment.html), as it's an updated version of the one you used for Spyder. If you switch over to that one, change the link near the top of the post.
         - I'm going to go through the tutorial tomorrow and will give you my recommendations for what to highlight here
         - Please add a section on what an "environment" is and how to set it up. @JorgeValde has some notes from Antonio about specifically how to set up an environment for this project; Jorge, please add them in a comment here when you have the instructions refined so that @Kirs10-Riley can add them to this tutorial. This new section should probably go ahead of, or at least early on in, the Spyder section.
      - If we put you on the GPU, you'll likely have to submit from the terminal, so keep that part but put it lower

**Reply by kbuzard** [reply]
@Kirs10-Riley Here's what I think our documentation should say in the Spyder section (you can pretty it up)
- Read and work through [Spyder: The Scientific Python Development Environment](https://fangohr.github.io/blog/spyder-the-scientific-python-development-environment.html)
   - Skip [Section 1.4.3](https://fangohr.github.io/blog/spyder-the-scientific-python-development-environment.html#automatic-symbolic-python) if you don't have a LaTeX installation.
   - In Section 1.6, replace "Run --> Configure or by pressing F6" with "Run --> Configuration per file or by pressing Ctrl + F6"
   - Just read through Sections 1.8 and 1.9; come back later when you're actually writing or debugging a program to work through these sections.
